src/lib.rs

```diff
@@ -0,0 +1,398 @@
+//! A `cargo` subcommand for the client-side Web.
+
+#![deny(
+    // missing_debug_implementations,
+    trivial_numeric_casts,
+    unstable_features,
+    unused_import_braces,
+    unused_qualifications
+)]
+
+#[macro_use]
+extern crate structopt;
+extern crate clap;
+extern crate digest;
+extern crate futures;
+extern crate http;
+extern crate hyper;
+extern crate libflate;
+extern crate notify;
+extern crate pbr;
+extern crate reqwest;
+extern crate serde;
+extern crate sha1;
+extern crate sha2;
+extern crate tar;
+extern crate tempfile;
+extern crate toml;
+#[macro_use]
+extern crate serde_derive;
+#[macro_use]
+extern crate serde_json;
+extern crate base_x;
+extern crate handlebars;
+extern crate indexmap;
+extern crate regex;
+extern crate unicode_categories;
+extern crate walkdir;
+extern crate websocket;
+#[macro_use]
+extern crate lazy_static;
+extern crate directories;
+extern crate percent_encoding;
+
+extern crate parity_wasm;
+#[macro_use]
+extern crate log;
+extern crate env_logger;
+extern crate rustc_demangle;
+
+extern crate ansi_term;
+extern crate cargo_metadata;
+
+extern crate memmap;
+extern crate semver;
+
+extern crate atty;
+extern crate open;
+#[macro_use]
+extern crate failure;
+
+mod cargo_shim;
+
+#[macro_use]
+mod utils;
+mod build;
+mod chrome_devtools;
+mod cmd_build;
+mod cmd_deploy;
+mod cmd_prepare_emscripten;
+mod cmd_start;
+mod cmd_test;
+mod config;
+mod deployment;
+mod emscripten;
+mod error;
+mod http_utils;
+mod package;
+mod project_dirs;
+mod test_chromium;
+mod wasm;
+mod wasm_context;
+mod wasm_export_main;
+mod wasm_export_table;
+mod wasm_gc;
+mod wasm_hook_grow;
+mod wasm_inline_js;
+mod wasm_intrinsics;
+mod wasm_js_export;
+mod wasm_js_snippet;
+mod wasm_runtime;
+
+use std::ffi::OsStr;
+use std::net::IpAddr;
+use std::path::PathBuf;
+
+use build::{Backend, BuildArgs};
+use cargo_shim::MessageFormat;
+use error::Error;
+use wasm_runtime::RuntimeKind;
+
+/// CLI for `cargo-web`
+#[derive(Debug, StructOpt)]
+#[structopt(name = "cargo-web")]
+#[structopt(about = "A `cargo` subcommand for the client-side web.")]
+#[structopt(raw(global_setting = "structopt::clap::AppSettings::ColoredHelp"))]
+#[structopt(rename_all = "kebab-case")]
+pub enum CargoWeb {
+    /// Compile a local package and all of its dependencies
+    Build {
+        #[structopt(flatten)]
+        build_args: Build,
+        #[structopt(flatten)]
+        ext: BuildExt,
+        #[structopt(flatten)]
+        build_target: Target,
+    },
```

Comment:
Make all of these variants like this:

```
Build( BuildOpts ),
Check( CheckOpts )
```

Reply:
Just a style preference? It looks like this would add a level of nesting for each subcommand, and we would still need to make those structs `pub` if we want to allow library usage.

Reply:
It's mostly for two reasons:
1. If you want to pass around all arguments to a given subcommand to some function as-is you'd have to either pass the whole enum and then do a `match` inside, or manually unpack the fields from the enum variant; on the other hand if you have it written like I've suggested here then you can just pass the whole struct in a single go, and you can move the definition of the struct which defines these arguments to the same file where the command is defined.
2. Visibility, since enum variant fields are all public-by-default, and I'd prefer if the fields here were a little bit more encapsulated. That is - I don't want the user of the library to have direct access to `Build` and `BuildExt`.

Reply:
Fair point on the first one - i'll go ahead and splice those out as structs. That will additionally simplify the fn signatures of `cmd_*::command_*`.

However, the second point is tough to reconcile - to programmatically construct a run configuration, a hypothetical user will need access to those settings (or just always use the `Default`, which weakens the argument for exposing this API).

Reply:
> However, the second point is tough to reconcile - to programmatically construct a run configuration, a hypothetical user will need access to those settings (or just always use the `Default`, which weakens the argument for exposing this API).

Hmm... yes, this is a good point. However, for now we could take advantage of `StructOpt` here, couldn't we? E.g. `BuildOpts` is going to implement the `StructOpt` trait, so the user can simply use `StructOpt::from_iter` to construct the `BuildOpts` and since our CLI is stable this usage will be stable too.

I agree that having this exposed as a real struct would be potentially nicer, but that *is* a compatibility hazard, so I think doing it this way will just be simpler for now while still exposing all of the functionality, and we can always potentially add another interface later.